### PR TITLE
Multi-panel polygon masking in dials.image_viewer

### LINKED ISSUE #2712: Mask generation for multi-panel detectors
How should masks for multi-panel detectors be generated?

Our new XFEL detector (CITIUS) has 72 panels. My pipeline masks module borders and stores it in the NXmx file but users need to add masks for the beam stopper, the injector nozzle etc. Because they vary from experiment to experiment,  they cannot be hard-coded in the preprocessing pipeline.

`dials.image_viewer`'s masking tool complains "Point must be within a panel" (circle tool), "All points must be contained within a single panel" (polygon tool).

`dials.generate_mask imported.expt circle=-23,5,100` (coordinates were read from the image viewer's toolbar) made a mask on the first panel (the top left panel in the image), not on the image center.

(The beam stopper for our previous detector, MPCCD, was glued to the capton window and lacked the "arm". Thus, we could use the `d_max` option for masking. This does not apply to CITIUS.)

@phyy-nx do you have any experiences with CSPAD?

image

### PR DESCRIPTION
**Pull request by biochem-fan**
This resolves #2712.

A polygon can be drawn over multiple panels. Vertices in panel gaps are also allowed. The resulting polygon is divided into sub-masks, each of which is contained within a single panel so that `dials.generate_mask` can handle it.

An example:
image

As pointed out by @dagewa in #2712, cyan lines run through the center of pixels. This is probably related to the way the coordinate system works in `dials.image_viewer` (#2190).

image

This does not pose problems; the resulting mask extends to the very edges of panels.

image

### CODE REVIEW

**Comment by phyy-nx** [comment]
Hi Takanori!  Very excited about this change, this comes up all the time.

I checked out the branch and tried it on a multipanel dataset.  I found a case where it throws an exception:
image

Here you can see several successful masks crossing several panels, but the central one, attempting to draw a mask on the four inner tiles, produced this exception:

```
Traceback (most recent call last):
  File "/dev/shm/aaron/modules/dials/src/dials/util/image_viewer/mask_frame.py", line 506, in OnUpdate
    self.AddUntrustedPolygon(self._mode_polygon_points)
  File "/dev/shm/aaron/modules/dials/src/dials/util/image_viewer/mask_frame.py", line 904, in AddUntrustedPolygon
    assert p_id == i
AssertionError
```

The image I used is in dials_data:
```
dials.image_viewer `dials.data get -q 4fluoro_cxi`/lcls_2022_smSFX_workshop_data/ten_cbfs/cxily6520_r0164_02081.cbf
```

What do you think?

**Comment by biochem-fan** [comment]
@phyy-nx Thank you very much for your report.
I reproduced your problem and is investigating.

**Comment by biochem-fan** [comment]
The difficulty is in the numerical fudges.

https://github.com/dials/dials/blob/2214232dd528799d3a8ac801fefa5fcd82a69570/src/dials/util/image_viewer/mask_frame.py#L894-L896

> This does not pose problems; the resulting mask extends to the very edges of panels.

I found that this does not hold always, depending on the panel configuration. I have to find a more robust approach.

**Comment by biochem-fan** [comment]
Investigating this, I found two issues in the underlying (single-panel) polygon masking code.

- It does not allow a triangle mask
- Some pixels are inclusive, while other pixels are exclusive

I will investigate and report them as separate issue(s). Meanwhile, I switch this PR to a draft state.

**Comment by biochem-fan** [comment]
> It does not allow a triangle mask

Reported to https://github.com/cctbx/dxtbx/issues/757.

**Comment by biochem-fan** [comment]
@phyy-nx I fixed the code. Could you please test again?

>> It does not allow a triangle mask

>  Reported to https://github.com/cctbx/dxtbx/issues/757.

I handled triangles in the image viewer code without changing dxtbx (for the moment).

> Some pixels are inclusive, while other pixels are exclusive

In [the polygon masking code in dxtbx](https://github.com/cctbx/dxtbx/blob/9e2674712261458c7828cf873d4582361e6c8165/src/dxtbx/masking/masking.h#L145-L147), an integer coordinate (i, j) means a corner of a pixel, while a pixel's center (i + 0.5, j + 0.5) must be within the polygon for the pixel to be masked. Thus, to fully mask a panel edge, we must go **one pixel beyond** the panel limit. That is, if a panel is 512 px wide, valid coordinates are from 0 to 511 (inclusive), but we have to specify 512 as the polygon vertex coordinate.

I'm not sure if this is user-friendly (especially with the 0.5 px error in the image viewer #2190), but I did not to change it in dxtbx for backward compatibility.

**Comment by phyy-nx** [comment]
Very nice

image

I agree about the half pixel thing, it's annoying, but doesn't need to be fixed here.

File: src/dials/util/image_viewer/mask_frame.py [hunk 1]
```diff
@@ -691,11 +694,11 @@ def OnLeftDown(self, event):
                 )
                 p1, p0, p_id = self._pyslip.tiles.flex_image.picture_to_readout(yc, xc)
 
-                if p_id < 0:
-                    return
+                # if p_id < 0:
+                #    return
```
Comment (phyy-nx) [inline src/dials/util/image_viewer/mask_frame.py:698]: Little bit of leftover clutter here, but otherwise LGTM

**Review by phyy-nx — APPROVED** [review]

File: src/dials/util/image_viewer/mask_frame.py (hunk 1 above)
Comment (biochem-fan) [reply src/dials/util/image_viewer/mask_frame.py:698]: Thanks! These lines and the next `if False` block are unnecessary. I will fix it.

**Pull request merged**